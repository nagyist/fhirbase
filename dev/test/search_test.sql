--db:fhirb
--{{{
\set org1 `cat test/fixtures/org1.json`
\set org_uuid '550e8400-e29b-41d4-a716-446655440009'
\set org_tags  '[{"scheme": "http://pt.com", "term": "http://pt/vip", "label":"pt"}]'

\set pt `cat test/fixtures/pt.json`
\set pt_uuid '550e8400-e29b-41d4-a716-446655440010'
\set pt_tags  '[{"scheme": "http://pt.com", "term": "http://pt/vip", "label":"pt"}]'

\set pt2 `cat test/fixtures/pt2.json`
\set pt2_uuid '550e8400-e29b-41d4-a716-446655440011'
\set pt2_tags  '[{"scheme": "http://pt.com", "term": "http://pt/noise", "label":"noise"}]'

\set doc_ref `cat test/fixtures/documentreference-example.json`
\set doc_ref_uuid '550e8400-e29b-41d4-a716-446655440012'

\set cfg '{"base":"https://test.me"}'

BEGIN;

INSERT into organization (logical_id, content) values (:'org_uuid', :'org1'::jsonb);
INSERT into patient (logical_id,content) values (:'pt_uuid',  format(:'pt', :'org_uuid')::jsonb);
INSERT into patient (logical_id,content) values (:'pt2_uuid', :'pt2'::jsonb);

SELECT assert_eq(
  :'pt_uuid',
  (SELECT string_agg(logical_id::text,' ') FROM search('Patient', 'name=roel')),
  'find pt');

SELECT assert_eq(
  :'pt_uuid',
  (SELECT string_agg(logical_id::text,' ') FROM search('Patient', 'identifier=123456789')),
  'search by identifier');

SELECT assert_eq(
  :'pt_uuid',
  (SELECT string_agg(logical_id::text,' ') FROM search('Patient', 'provider=' || :'org_uuid')),
  'search by provider');

SELECT assert_eq(
  :'pt_uuid',
  (SELECT string_agg(logical_id::text,' ') FROM search('Patient', 'provider=' || :'org_uuid')),
  'search by provider');

SELECT assert_eq( NULL ,
  (SELECT string_agg(logical_id::text,' ') FROM search('Patient', 'provider=nonexist')),
  'search by nonexisting provider');

SELECT assert_eq(2::bigint ,
  (SELECT count(*) FROM search('Patient', '')),
  'count patients');

SELECT assert_eq(1::bigint ,
  (SELECT count(*) FROM search('Patient', '_count=1')),
  'count patients with limit');

SELECT assert_eq(1::bigint ,
  (SELECT count(*) FROM search('Patient', format('_id=%s', :'pt_uuid'))),
  'search by id');

SELECT assert_eq(2::bigint ,
  (SELECT count(*) FROM search('Patient', format('_id=%s,%s', :'pt_uuid', :'pt2_uuid'))),
  'search by ids');

SELECT content#>'{name,0,given}' FROM search('Patient', '_sort=given');
SELECT content#>'{name,0,given}' FROM search('Patient', '_sort:desc=given');

SELECT content#>'{birthDate}' FROM search('Patient', '_sort=birthdate');
SELECT content#>'{birthDate}' FROM search('Patient', '_sort:desc=birthdate');

ROLLBACK;
--}}}
--{{{
select build_search_query('Patient', '_id=4');
select build_search_query('Patient', 'provider._id=4');
--}}}
