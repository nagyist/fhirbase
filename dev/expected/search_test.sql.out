SET escape_string_warning=off;
SET
BEGIN;
BEGIN
SELECT insert_resource('550e8400-e29b-41d4-a716-446655440009'::uuid,  E'{
  "resourceType": "Organization",
  "text": {
    "status": "generated",
    "div": "<div>\\n      Health Level Seven International<br/>\\n\\t\\t\\t\\t3300 Washtenaw Avenue, Suite 227<br/>\\n\\t\\t\\t\\tAnn Arbor, MI 48104<br/>\\n\\t\\t\\t\\tUSA<br/>\\n\\t\\t\\t\\t(+1) 734-677-7777 (phone)<br/>\\n\\t\\t\\t\\t(+1) 734-677-6622 (fax)<br/>\\n\\t\\t\\t\\tE-mail:  <a href=\\"mailto:hq@HL7.org\\">hq@HL7.org</a>\\n    </div>"
  },
  "name": "Health Level Seven International",
  "telecom": [
    {
      "system": "phone",
      "value": "(+1) 734-677-7777"
    },
    {
      "system": "fax",
      "value": "(+1) 734-677-6622"
    },
    {
      "system": "email",
      "value": "hq@HL7.org"
    }
  ],
  "address": [
    {
      "line": [
        "3300 Washtenaw Avenue, Suite 227"
      ],
      "city": "Ann Arbor",
      "state": "MI",
      "zip": "48104",
      "country": "USA"
    }
  ]
}'::jsonb, '[]'::jsonb);
           insert_resource            
--------------------------------------
 550e8400-e29b-41d4-a716-446655440009
(1 row)

SELECT insert_resource('550e8400-e29b-41d4-a716-446655440010'::uuid,  E'{
  "resourceType": "Patient",
  "text": {
    "status": "generated",
    "div": "<div>\\n      <p>\\n        <b>Generated Narrative</b>\\n      </p>\\n      <p>\\n        <b>identifier</b>: BSN = 123456789 (official), BSN = 123456789 (official)\\n      </p>\\n      <p>\\n        <b>name</b>: Roel(official)\\n      </p>\\n      <p>\\n        <b>telecom</b>: ph: +31612345678(mobile), ph: +31201234567(home)\\n      </p>\\n      <p>\\n        <b>gender</b>: \\n        <span title=\\"Codes: {http://snomed.info/sct 248153007}, {http://hl7.org/fhir/v3/AdministrativeGender M}\\">Male</span>\\n      </p>\\n      <p>\\n        <b>birthDate</b>: 13-Mar 1960\\n      </p>\\n      <p>\\n        <b>deceased[x]</b>: false\\n      </p>\\n      <p>\\n        <b>address</b>: Bos en Lommerplein 280 Amsterdam 1055RW NLD (home)\\n      </p>\\n      <p>\\n        <b>maritalStatus</b>: \\n        <span title=\\"Codes: {http://snomed.info/sct 36629006}, {http://hl7.org/fhir/v3/MaritalStatus M}\\">Legally married</span>\\n      </p>\\n      <p>\\n        <b>multipleBirth[x]</b>: false\\n      </p>\\n      <p>\\n        <b>photo</b>: \\n      </p>\\n      <h3>Contacts</h3>\\n      <table class=\\"grid\\">\\n        <tr>\\n          <td>\\n            <b>Relationship</b>\\n          </td>\\n          <td>\\n            <b>Name</b>\\n          </td>\\n          <td>\\n            <b>Telecom</b>\\n          </td>\\n          <td>\\n            <b>Address</b>\\n          </td>\\n          <td>\\n            <b>Gender</b>\\n          </td>\\n          <td>\\n            <b>Organization</b>\\n          </td>\\n        </tr>\\n        <tr>\\n          <td>\\n            <span title=\\"Codes: {http://snomed.info/sct 127850001}, {http://hl7.org/fhir/patient-contact-relationship partner}\\">Wife</span>\\n          </td>\\n          <td>Ariadne Bor-Jansma</td>\\n          <td>ph: +31201234567(home)</td>\\n          <td> </td>\\n          <td> </td>\\n          <td> </td>\\n        </tr>\\n      </table>\\n      <p>\\n        <b>communication</b>: \\n        <span title=\\"Codes: {urn:std:iso:639-1 nl-NL}\\">Dutch</span>\\n      </p>\\n      <p>\\n        <b>managingOrganization</b>: AUMC\\n      </p>\\n      <p>\\n        <b>active</b>: true\\n      </p>\\n    </div>"
  },
  "identifier": [
    {
      "use": "official",
      "label": "MRN",
      "system": "MRN",
      "value": "7777777"
    },
    {
      "use": "official",
      "label": "BSN",
      "system": "urn:oid:2.16.840.1.113883.2.4.6.3",
      "value": "123456789"
    }
  ],
  "name": [
    {
      "use": "official",
      "text": "Roel",
      "family": [
        "Bor"
      ],
      "given": [
        "Roelof Olaf"
      ],
      "prefix": [
        "Drs."
      ],
      "suffix": [
        "PDEng."
      ]
    }
  ],
  "telecom": [
    {
      "system": "phone",
      "value": "+31612345678",
      "use": "mobile"
    },
    {
      "system": "phone",
      "value": "+31201234567",
      "use": "home"
    }
  ],
  "gender": {
    "coding": [
      {
        "system": "http://snomed.info/sct",
        "code": "248153007",
        "display": "Male"
      },
      {
        "system": "http://hl7.org/fhir/v3/AdministrativeGender",
        "code": "M",
        "display": "Male"
      }
    ]
  },
  "birthDate": "1960-03-13",
  "deceasedBoolean": false,
  "address": [
    {
      "use": "home",
      "line": [
        "Bos en Lommerplein 280"
      ],
      "city": "Amsterdam",
      "zip": "1055RW",
      "country": "NLD"
    }
  ],
  "maritalStatus": {
    "coding": [
      {
        "system": "http://snomed.info/sct",
        "code": "36629006",
        "display": "Legally married"
      },
      {
        "system": "http://hl7.org/fhir/v3/MaritalStatus",
        "code": "M"
      }
    ]
  },
  "multipleBirthBoolean": false,
  "photo": [
    {
      "contentType": "image/jpeg",
      "url": "binary/@f006"
    }
  ],
  "contact": [
    {
      "relationship": [
        {
          "coding": [
            {
              "system": "http://snomed.info/sct",
              "code": "127850001",
              "display": "Wife"
            },
            {
              "system": "http://hl7.org/fhir/patient-contact-relationship",
              "code": "partner"
            }
          ]
        }
      ],
      "name": {
        "use": "usual",
        "text": "Ariadne Bor-Jansma"
      },
      "telecom": [
        {
          "system": "phone",
          "value": "+31201234567",
          "use": "home"
        }
      ]
    }
  ],
  "communication": [
    {
      "coding": [
        {
          "system": "urn:std:iso:639-1",
          "code": "nl-NL",
          "display": "Dutch"
        }
      ]
    }
  ],
  "managingOrganization": {
    "reference": "Organization/550e8400-e29b-41d4-a716-446655440009",
    "display": "AUMC"
  },
  "active": true
}'::jsonb, '[{"scheme": "http://pt.com", "term": "http://pt/vip", "label":"pt"}]'::jsonb);
           insert_resource            
--------------------------------------
 550e8400-e29b-41d4-a716-446655440010
(1 row)

SELECT insert_resource('550e8400-e29b-41d4-a716-446655440011'::uuid,  E'{
  "resourceType": "Patient",
  "text": {
    "status": "generated",
    "div": "<div>\\n      <table>\\n        <tbody>\\n          <tr>\\n            <td>Name</td>\\n            <td>Peter James <b>Chalmers</b> (&quot;Jim&quot;)</td>\\n          </tr>\\n          <tr>\\n            <td>Address</td>\\n            <td>534 Erewhon, Pleasantville, Vic, 3999</td>\\n          </tr>\\n          <tr>\\n            <td>Contacts</td>\\n            <td>Home: unknown. Work: (03) 5555 6473</td>\\n          </tr>\\n          <tr>\\n            <td>Id</td>\\n            <td>MRN: 12345 (Acme Healthcare)</td>\\n          </tr>\\n        </tbody>\\n      </table>\\n    </div>"
  },
  "identifier": [
    {
      "use": "usual",
      "label": "MRN",
      "system": "urn:oid:1.2.36.146.595.217.0.1",
      "value": "12345",
      "period": {
        "start": "2001-05-06"
      },
      "assigner": {
        "display": "Acme Healthcare"
      }
    }
  ],
  "name": [
    {
      "use": "official",
      "family": [
        "Chalmers"
      ],
      "given": [
        "Peter",
        "James"
      ]
    },
    {
      "use": "usual",
      "given": [
        "Jim"
      ]
    }
  ],
  "gender": {
    "coding": [
      {
        "system": "http://hl7.org/fhir/v3/AdministrativeGender",
        "code": "M",
        "display": "Male"
      }
    ]
  },
  "birthDate": "1974-12-25",
  "deceasedBoolean": false,
  "address": [
    {
      "use": "home",
      "line": [
        "534 Erewhon St"
      ],
      "city": "PleasantVille",
      "state": "Vic",
      "zip": "3999"
    }
  ],
  "contact": [
    {
      "relationship": [
        {
          "coding": [
            {
              "system": "http://hl7.org/fhir/patient-contact-relationship",
              "code": "partner"
            }
          ]
        }
      ],
      "name": {
        "family": [
          "du",
          "Marché"
        ],
        "_family": [
          {
            "extension": [
              {
                "url": "http://hl7.org/fhir/Profile/iso-21090#qualifier",
                "valueCode": "VV"
              }
            ]
          },
          null
        ],
        "given": [
          "Bénédicte"
        ]
      },
      "telecom": [
        {
          "system": "phone",
          "value": "+33 (237) 998327"
        }
      ]
    }
  ],
  "managingOrganization": {
    "reference": "Organization/1"
  },
  "active": false
}'::jsonb, '[{"scheme": "http://pt.com", "term": "http://pt/noise", "label":"noise"}]'::jsonb);
           insert_resource            
--------------------------------------
 550e8400-e29b-41d4-a716-446655440011
(1 row)

SELECT insert_resource('550e8400-e29b-41d4-a716-446655440012'::uuid,  E'{"resourceType":"DocumentReference","text":{"status":"generated","div":"<div>\\n      <p>\\n        <b>Generated Narrative</b>\\n      </p>\\n      <p>\\n        <b>masterIdentifier</b>: urn:oid:1.3.6.1.4.1.21367.2005.3.7\\n      </p>\\n      <p>\\n        <b>subject</b>: \\n        <a href=\\"patient-example-xcda.html\\">MRN = 12345 (usual); Henry Levin ; Male; birthDate: 24-Sep 1932; active</a>\\n      </p>\\n      <p>\\n        <b>type</b>: \\n        <span title=\\"Codes: {http://loinc.org 34108-1}\\">Outpatient Note</span>\\n      </p>\\n      <p>\\n        <b>author</b>: Sherry Dopplemeyer ; Primary Surgon; Orthopedic, Gerald Smitty ; Attending; Orthopedic\\n      </p>\\n      <p>\\n        <b>created</b>: 24-Dec 2005 9:35\\n      </p>\\n      <p>\\n        <b>indexed</b>: 24-Dec 2005 9:43\\n      </p>\\n      <p>\\n        <b>status</b>: current\\n      </p>\\n      <p>\\n        <b>description</b>: Physical\\n      </p>\\n      <p>\\n        <b>confidentiality</b>: \\n        <span title=\\"Codes: {http://ihe.net/xds/connectathon/confidentialityCodes 1.3.6.1.4.1.21367.2006.7.101}\\">Clinical-Staff</span>\\n      </p>\\n      <p>\\n        <b>primaryLanguage</b>: en-US\\n      </p>\\n      <p>\\n        <b>mimeType</b>: application/hl7-v3+xml\\n      </p>\\n      <p>\\n        <b>size</b>: 3654\\n      </p>\\n      <p>\\n        <b>hash</b>: da39a3ee5e6b4b0d3255bfef95601890afd80709\\n      </p>\\n      <p>\\n        <b>location</b>: \\n        <a href=\\"http://example.org/xds/mhd/Binary/@07a6483f-732b-461e-86b6-edb665c45510\\">http://example.org/xds/mhd/Binary/@07a6483f-732b-461e-86b6-edb665c45510</a>\\n      </p>\\n      <h3>Contexts</h3>\\n      <table class=\\"grid\\">\\n        <tr>\\n          <td>\\n            <b>Event</b>\\n          </td>\\n          <td>\\n            <b>Period</b>\\n          </td>\\n          <td>\\n            <b>FacilityType</b>\\n          </td>\\n        </tr>\\n        <tr>\\n          <td>\\n            <span title=\\"Codes: {http://ihe.net/xds/connectathon/eventCodes T-D8200}\\">Arm</span>\\n          </td>\\n          <td>23-Dec 2004 8:0 --&gt; 23-Dec 2004 8:1</td>\\n          <td>\\n            <span title=\\"Codes: {http://www.ihe.net/xds/connectathon/healthcareFacilityTypeCodes Outpatient}\\">Outpatient</span>\\n          </td>\\n        </tr>\\n      </table>\\n    </div>"},"contained":[{"resourceType":"Practitioner","_id":"a1","name":{"family":["Dopplemeyer"],"given":["Sherry"]},"telecom":[{"system":"email","value":"john.doe@healthcare.example.org"}],"organization":{"display":"Cleveland Clinic"},"role":[{"text":"Primary Surgon"}],"specialty":[{"text":"Orthopedic"}]},{"resourceType":"Practitioner","_id":"a2","name":{"family":["Smitty"],"given":["Gerald"]},"telecom":[{"system":"email","value":"john.doe@healthcare.example.org"}],"organization":{"display":"Cleveland Clinic"},"role":[{"text":"Attending"}],"specialty":[{"text":"Orthopedic"}]}],"masterIdentifier":{"system":"urn:ietf:rfc:3986","value":"urn:oid:1.3.6.1.4.1.21367.2005.3.7"},"subject":{"reference":"Patient/xcda"},"type":{"coding":[{"system":"http://loinc.org","code":"34108-1","display":"Outpatient Note"}]},"author":[{"reference":"#a1"},{"reference":"#a2"}],"created":"2005-12-24T09:35:00+11:00","indexed":"2005-12-24T09:43:41+11:00","status":"current","description":"Physical","confidentiality":[{"coding":[{"system":"http://ihe.net/xds/connectathon/confidentialityCodes","code":"1.3.6.1.4.1.21367.2006.7.101","display":"Clinical-Staff"}]}],"primaryLanguage":"en-US","mimeType":"application/hl7-v3+xml","size":3654,"hash":"da39a3ee5e6b4b0d3255bfef95601890afd80709","location":"http://example.org/xds/mhd/Binary/@07a6483f-732b-461e-86b6-edb665c45510","context":{"event":[{"coding":[{"system":"http://ihe.net/xds/connectathon/eventCodes","code":"T-D8200","display":"Arm"}]}],"period":{"start":"2004-12-23T08:00:00","end":"2004-12-23T08:01:00"},"facilityType":{"coding":[{"system":"http://www.ihe.net/xds/connectathon/healthcareFacilityTypeCodes","code":"Outpatient","display":"Outpatient"}]}}}'::jsonb, '[]'::jsonb);
           insert_resource            
--------------------------------------
 550e8400-e29b-41d4-a716-446655440012
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010', logical_id, 'pt found by name')
  FROM search('Patient', 'name=roel');
      assert_eq      
---------------------
 OK pt found by name
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010', logical_id, 'pt found by identifier')
  FROM search('Patient', 'identifier=123456789');
         assert_eq         
---------------------------
 OK pt found by identifier
(1 row)

SELECT build_search_query('Patient', _parse_param('identifier=MRN|7777777'));
                                                              build_search_query                                                               
-----------------------------------------------------------------------------------------------------------------------------------------------
                                                                                                                                              +
     SELECT ROW_NUMBER() OVER () as weight, x.*                                                                                               +
       FROM (                                                                                                                                 +
             SELECT DISTINCT patient.category                                                                                                 +
 ,patient.content                                                                                                                             +
 ,patient.logical_id                                                                                                                          +
 ,patient.published                                                                                                                           +
 ,patient.resource_type                                                                                                                       +
 ,patient.updated                                                                                                                             +
 ,patient.version_id                                                                                                                          +
               FROM patient patient                                                                                                           +
                                                                                                                                              +
             JOIN patient_search_token patient_identifier                                                                                     +
               ON patient_identifier.resource_id::varchar = patient.logical_id::varchar                                                       +
              AND (patient_identifier.param = 'identifier' AND (patient_identifier.code = '7777777' AND patient_identifier.namespace = 'MRN'))+
                                                                                                                                              +
               WHERE  true = true                                                                                                             +
               ORDER BY  updated DESC                                                                                                         +
               LIMIT 100                                                                                                                      +
               OFFSET 0                                                                                                                       +
             ) x                                                                                                                              +
   
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010',
 (SELECT logical_id
    FROM search('Patient', 'identifier=MRN|7777777'))
 ,'pt found by mrn');
     assert_eq      
--------------------
 OK pt found by mrn
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010', logical_id, 'pt found by status')
  FROM search('Patient', 'active=true');
       assert_eq       
-----------------------
 OK pt found by status
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010', logical_id, 'pt found by status')
  FROM search('Patient', 'active=true');
       assert_eq       
-----------------------
 OK pt found by status
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010',
 (SELECT logical_id
    FROM search('Patient', 'telecom=+31612345678'))
 ,'pt found by phone');
      assert_eq       
----------------------
 OK pt found by phone
(1 row)

SELECT build_search_query('Patient', _parse_param('telecom:missing=true'));
                                   build_search_query                                   
----------------------------------------------------------------------------------------
                                                                                       +
     SELECT ROW_NUMBER() OVER () as weight, x.*                                        +
       FROM (                                                                          +
             SELECT DISTINCT patient.category                                          +
 ,patient.content                                                                      +
 ,patient.logical_id                                                                   +
 ,patient.published                                                                    +
 ,patient.resource_type                                                                +
 ,patient.updated                                                                      +
 ,patient.version_id                                                                   +
               FROM patient patient                                                    +
                                                                                       +
              LEFT JOIN patient_search_string patient_string                           +
                   ON patient_string.resource_id::varchar = patient.logical_id::varchar+
                  AND patient_string.param = 'telecom'                                 +
                                                                                       +
               WHERE patient_string.resource_id IS NULL                                +
               ORDER BY  updated DESC                                                  +
               LIMIT 100                                                               +
               OFFSET 0                                                                +
             ) x                                                                       +
   
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440011',
 (SELECT string_agg(logical_id::text,'|')
    FROM search('Patient', 'telecom:missing=true'))
 ,'test missing true');
      assert_eq       
----------------------
 OK test missing true
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010',
 (SELECT string_agg(logical_id::text,'|')
    FROM search('Patient', 'telecom:missing=false'))
 ,'test missing false');
       assert_eq       
-----------------------
 OK test missing false
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010',
 (SELECT logical_id
    FROM search('Patient', 'gender=http://snomed.info/sct|248153007'))
 ,'pt found by snomed gender');
          assert_eq           
------------------------------
 OK pt found by snomed gender
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010',
 (SELECT logical_id
    FROM search('Patient', 'birthdate=1960'))
 ,'pt found birthdate');
       assert_eq       
-----------------------
 OK pt found birthdate
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440009',
 (SELECT logical_id
    FROM search('Organization', 'name=Health Level'))
 ,'org by name');
   assert_eq    
----------------
 OK org by name
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010',
 (SELECT logical_id
    FROM search('Patient', 'provider.name=Health%20Level') LIMIT 1)
 ,'pt by org name');
     assert_eq     
-------------------
 OK pt by org name
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010',
 (SELECT logical_id
    FROM search('Patient', 'name=Roelof&provider.name=Health%20Level'))
 ,'pt by name & org name');
        assert_eq         
--------------------------
 OK pt by name & org name
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010',
 (SELECT logical_id
    FROM search('Patient', 'family=bor') LIMIT 1)
 ,'pt by family');
    assert_eq    
-----------------
 OK pt by family
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440010',
 (SELECT logical_id
    FROM search('Patient', 'family=bor&name=bor') LIMIT 1)
 ,'pt by family & name');
       assert_eq        
------------------------
 OK pt by family & name
(1 row)

SELECT assert_eq('http://pt/vip',
 (SELECT string_agg(jsonb_array_elements#>>'{category,0,term}','')
    FROM jsonb_array_elements(
            fhir_search('{"base":"https://test.me"}'::jsonb,
              'Patient'::varchar, '_tag=http%3A%2F%2Fpt%2Fvip')->'entry'))
 ,'pt by tag');
  assert_eq   
--------------
 OK pt by tag
(1 row)

SELECT assert_eq('550e8400-e29b-41d4-a716-446655440012',
   (SELECT string_agg(logical_id::varchar, '|')),
  'search number')
  FROM search('DocumentReference', 'size=>100');
    assert_eq     
------------------
 OK search number
(1 row)

SELECT assert_eq(NULL,
   (SELECT string_agg(logical_id::varchar, '|')),
  'search number')
  FROM search('DocumentReference', 'size=<100');
    assert_eq     
------------------
 OK search number
(1 row)

ROLLBACK;
ROLLBACK
